### PR TITLE
[SPARK-11191] [SPARK-11311] [SQL] Backports #9664 and #9277 to branch-1.5

### PR DESCRIPTION
**Pull request by liancheng**
The main purpose of this PR is to backport #9664, which depends on #9277.

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala [hunk 1]
```diff
@@ -57,9 +57,11 @@ import org.apache.spark.util.Utils
 /**
  * This is the HiveQL Dialect, this dialect is strongly bind with HiveContext
  */
-private[hive] class HiveQLDialect extends ParserDialect {
+private[hive] class HiveQLDialect(sqlContext: HiveContext) extends ParserDialect {
   override def parse(sqlText: String): LogicalPlan = {
-    HiveQl.parseSql(sqlText)
+    sqlContext.executionHive.withHiveState {
+      HiveQl.parseSql(sqlText)
+    }
```
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala]: `HiveQl` tries to resolve UDTFs using Hive `FunctionRegistry` during parsing, must wrap it with execution Hive state.

**Comment by liancheng** [comment]
Hm, I suspect maybe the Jenkins build was executed before #9570 (the ADD JAR issue) was merged to branch-1.5, and thus caused the failure.

**Comment by liancheng** [comment]
retest this please

**Comment by liancheng** [comment]
The test failure was because a previous test case (`HiveThriftBinaryServerSuite.test jdbc cancel`) sets `spark.sql.hive.thriftServer.async` to `false`. Haven't looked into why this fails ADD JAR, but I think we can address this issue in a separate PR.

**Comment by liancheng** [comment]
cc @yhuai

**Comment by yhuai** [comment]
Merge to branch 1.5.

**Comment by yhuai** [comment]
Merged.

**Pull request closed without merging**